source/simulator/assemblers/newton_stokes.cc

```diff
@@ -630,6 +632,424 @@ namespace aspect
+    {
+      internal::Assembly::Scratch::StokesSystem<dim> &scratch = dynamic_cast<internal::Assembly::Scratch::StokesSystem<dim>&> (scratch_base);
+      internal::Assembly::CopyData::StokesSystem<dim> &data = dynamic_cast<internal::Assembly::CopyData::StokesSystem<dim>&> (data_base);
+
+      const Introspection<dim> &introspection = this->introspection();
+      const FiniteElement<dim> &fe = this->get_fe();
+      const unsigned int stokes_dofs_per_cell = data.local_dof_indices.size();
+      const unsigned int n_q_points           = scratch.finite_element_values.n_quadrature_points;
+      const double derivative_scaling_factor = this->get_newton_handler().parameters.newton_derivative_scaling_factor;
+      const double pressure_scaling = this->get_pressure_scaling();
+
+      AssertThrow(dynamic_cast<const MaterialModel::ViscoPlasticSimple<dim> *>(&this->get_material_model()), ExcInternalError());
+
+      const bool enable_additional_stokes_rhs = this->get_parameters().enable_additional_stokes_rhs;
+      const MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> *force = 
+        enable_additional_stokes_rhs
+        ? scratch.material_model_outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim>>()
+        : nullptr;
+
+      const bool enable_elasticity = this->get_parameters().enable_elasticity;
+      const MaterialModel::ElasticOutputs<dim> *elastic_out =
+        scratch.material_model_outputs.template get_additional_output<MaterialModel::ElasticOutputs<dim>>();
+      if (enable_elasticity)
+        AssertThrow(elastic_out != nullptr,
+                    ExcMessage("Error: The Newton method requires ElasticOutputs when elasticity is enabled."));
+
+      std::vector<SymmetricTensor<2,dim>> effective_strain_rates(n_q_points);
+
+      // always compute the common terms in the weak form of Stokes equation
+      for (unsigned int q = 0; q < n_q_points; ++q)
+        {
+          for (unsigned int i = 0, i_stokes = 0; i_stokes < stokes_dofs_per_cell; /*increment at end of loop*/)
+            {
+              if (introspection.is_stokes_component(fe.system_to_component_index(i).first))
+                {
+                  scratch.phi_u[i_stokes] = scratch.finite_element_values[introspection.extractors.velocities].value (i, q);
+                  scratch.phi_p[i_stokes] = scratch.finite_element_values[introspection.extractors.pressure].value (i, q);
+                  scratch.grads_phi_u[i_stokes] = scratch.finite_element_values[introspection.extractors.velocities].symmetric_gradient(i, q);
+                  scratch.div_phi_u[i_stokes]   = scratch.finite_element_values[introspection.extractors.velocities].divergence (i, q);
+
+#if DEBUG
+                  // This is needed to test the velocity part of the matrix for
+                  // being symmetric positive-definite.
+                  scratch.dof_component_indices[i_stokes] = fe.system_to_component_index(i).first;
+#endif
```

Comment:
I'm not in favor of having code only in debug mode, because it means that you can't use this array in release mode somewhere else. Is this done in other places already? If so, we should probably fix those as well.